### PR TITLE
Readability

### PR DESCRIPTION
**Pull request by FG-TUM**
# Description

- [x] Some renaming of variables 
- [x] more verbose documentation. 
- [x] ~No actual code changes.~  -> Some changes to simplify code
- [x] Minimal changes to improve performance (e.g. `reserve` before `push_back`)
- [x] Add time measurement of rebuild neighbor lists
- [x] VLC: call reserve on every neighbor list to drastically speed up neighbor list rebuild (factor >2.1) at the cost of slightly slower iteratePairwise (~5%)

### CODE REVIEW

**Comment by SteffenSeckler** [comment]
please also remove `getAllowedConfigurations` from the AutoTuner. It's not even implemented...

**Comment by FG-TUM** [comment]
> please also remove `getAllowedConfigurations` from the AutoTuner. It's not even implemented...

Done in 79329ce

File: src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h [hunk 1]
```diff
@@ -240,45 +241,54 @@ class VerletClusterListsRebuilder {
...
-        // calculate distance in xy-plane and skip if already longer than interactionLength
+        // calculate distance in xy-plane
         auto distBetweenTowersXYsqr = distBetweenTowersX * distBetweenTowersX + distBetweenTowersY * distBetweenTowersY;
+        // skip if already longer than interactionLength
         if (distBetweenTowersXYsqr <= _interactionLengthSqr) {
-          auto &neighborTower = getTowerAtCoordinates(neighborIndexX, neighborIndexY);
+          auto &neighborTower = getTower(neighborIndexX, neighborIndexY);
 
-          calculateNeighborsForTowerPair(tower, neighborTower, distBetweenTowersXYsqr, useNewton3);
+          (this->*function)(tower, neighborTower, distBetweenTowersXYsqr, useNewton3);
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h:291]: I don't like this method of passing member functions.
Best solution would be to make the passed function static.

File: src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h [hunk 2]
```diff
@@ -301,66 +311,72 @@ class VerletClusterListsRebuilder {
...
    *
-   * @param towerIndexX The x-coordinate of the given tower.
-   * @param towerIndexY The y-coordinate of the given tower.
-   * @param neighborIndexX The x-coordinate of the given neighbor tower.
-   * @param neighborIndexY The y-coordinate of the given neighbor tower.
+   * @param towerIndexX The x-index of the given tower.
+   * @param towerIndexY The y-index of the given tower.
+   * @param neighborIndexX The x-index of the given neighbor tower.
+   * @param neighborIndexY The y-index of the given neighbor tower.
    * @return True, if clusters of the given tower should contain clusters of the given neighbor tower as neighbors with
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h:324]: documentation of return should be adapted.

File: src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h [hunk 3]
```diff
@@ -301,66 +311,72 @@ class VerletClusterListsRebuilder {
...
-
-      if (towerRealCluster) {
-        for (size_t neighborIndex = startIndexNeighbor; neighborIndex < neighborTower.getNumClusters();
-             neighborIndex++) {
-          const bool isSameCluster = towerIndex == neighborIndex;
-          if (not useNewton3 and isSameTower and isSameCluster) {
+  void calculateNeighborsBetweenTowers(internal::ClusterTower<Particle> &towerA,
+                                       internal::ClusterTower<Particle> &towerB, double distBetweenTowersXYsqr,
+                                       bool useNewton3) {
+    const bool isSameTower = &towerA == &towerB;
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h:359]:
```suggestion
    const bool isSameTower = (&towerA == &towerB);
```
for readability

File: src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h [hunk 4]
```diff
@@ -301,66 +311,72 @@ class VerletClusterListsRebuilder {
...
+    for (size_t clusterIndexInTowerA = 0; clusterIndexInTowerA < towerA.getNumClusters(); clusterIndexInTowerA++) {
+      // if we are within one tower depending on newton3 only look at forward neighbors
+      auto startClusterIndexInTowerB = isSameTower and useNewton3 ? clusterIndexInTowerA + 1 : 0;
+      auto &clusterA = towerA.getCluster(clusterIndexInTowerA);
+      auto [clusterABoxBottom, clusterABoxTop, clusterAContainsParticles] = clusterA.getZMinMax();
+
+      if (clusterAContainsParticles) {
+        for (size_t clusterIndexInTowerB = startClusterIndexInTowerB; clusterIndexInTowerB < towerB.getNumClusters();
+             clusterIndexInTowerB++) {
+          // a cluster can not be a neighbor to itself
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h:369]:
```suggestion
          // a cluster cannot be a neighbor to itself
```

File: src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h [hunk 5]
```diff
@@ -301,66 +311,72 @@ class VerletClusterListsRebuilder {
...
+          // when newton3 is true this is not possible since the choice of the start index
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h:370]: sentence is not complete.
Maybe this was meant:
```suggestion
          // If newton3 is true this is not possible because of the choice of the start index.
```

File: src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h [hunk 6]
```diff
@@ -151,75 +153,89 @@ class VerletLists
...
+        _soaNeighborLists[i_id][j] = _particlePtr2indexMap[neighborPtr];
         j++;
       }
     }
-    AutoPasLog(debug,
-               "VerletLists::generateSoAListFromAoSVerletLists: average verlet list "
-               "size is {}",
-               static_cast<double>(accumulatedListSize) / _aosNeighborLists.size());
+
+    if (autopas::Logger::get()->level() <= autopas::Logger::LogLevel::debug) {
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletLists/VerletLists.h:204]: why did you add this if here?

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h [hunk 7]
```diff
@@ -6,7 +6,7 @@
 
 #pragma once
 
-#include "VerletListsCellsHelpers.h"
+#include "autopas/containers//verletListsCellBased/verletListsCells/VerletListsCellsHelpers.h"
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h:9]:
```suggestion
#include "autopas/containers/verletListsCellBased/verletListsCells/VerletListsCellsHelpers.h"
```
one `/` too many.
It's actually in the same directory, so not pretending the path is ok.

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h [hunk 8]
```diff
@@ -21,6 +21,7 @@
 #include "autopas/options/DataLayoutOption.h"
 #include "autopas/options/LoadEstimatorOption.h"
 #include "autopas/options/TraversalOption.h"
+#include "autopas/selectors/TraversalSelector.h"
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h:24]: check if this does not produce a circular dependency

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h [hunk 9]
```diff
@@ -121,69 +123,47 @@ class VerletListsCells
...
+    const auto indices = _particleToCellMap.at(const_cast<Particle *>(particle));
     return _neighborLists.at(indices.first).at(indices.second).second;
   }
 
   void rebuildNeighborLists(TraversalInterface *traversal) override {
     bool useNewton3 = traversal->getUseNewton3();
     this->_verletBuiltNewton3 = useNewton3;
 
-    // create a Verlet Lists for each cell
+    // initialize a Verlet Lists for each cell
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h:134]:
```suggestion
    // Initialize a neighbor list for each cell.
```

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h [hunk 10]
```diff
@@ -121,69 +123,47 @@ class VerletListsCells
...
+      _neighborLists[cellIndex].reserve(cells[cellIndex].numParticles());
+      size_t particleIndexWithinCell = 0;
+      for (auto iter = cells[cellIndex].begin(); iter.isValid(); ++iter, ++particleIndexWithinCell) {
         Particle *particle = &*iter;
-        _neighborLists[cellIndex].push_back(
-            std::pair<Particle *, std::vector<Particle *>>(particle, std::vector<Particle *>()));
-        _cellMap[particle] = std::pair<size_t, size_t>(cellIndex, i);
+        _neighborLists[cellIndex].emplace_back(particle, std::vector<Particle *>());
+        // reserve space for 5 times as many particles in the list as are in the particles' cell
+        // 5 is an empirically determined magic number
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h:146]:
```suggestion
        // In a cell with N particles, reserve space for 5N neighbors.
        // 5 is an empirically determined magic number that provides good speed.
```

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h [hunk 11]
```diff
@@ -121,69 +123,47 @@ class VerletListsCells
...
-      }
-      default:
-        utils::ExceptionHandler::exception("VerletListsCells::updateVerletLists(): unsupported Traversal: {}",
-                                           _buildTraversal);
-        break;
-    }
+    typename VerletListsCellsHelpers<Particle>::VerletListGeneratorFunctor f(_neighborLists, _particleToCellMap,
+                                                                             this->getCutoff() + this->getSkin());
+
+    // generate the build traversal with the traversal selector and apply the build functor with it
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h:155]:
```suggestion
    // Generate the build traversal with the traversal selector and apply the build functor with it.
```

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h [hunk 12]
```diff
@@ -121,69 +123,47 @@ class VerletListsCells
...
+    TraversalSelector<LinkedParticleCell> traversalSelector;
+    // argument "cluster size" does not matter here
```
Comment (SteffenSeckler) [inline src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h:157]:
```suggestion
    // Argument "cluster size" does not matter here.
```

File: src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h (hunk 8 above)
Comment (FG-TUM) [reply src/autopas/containers/verletListsCellBased/verletListsCells/VerletListsCells.h:24]: Nope, we are fine there since `TraversalSelector` only includes traversals and no Containers. Also, the traversals do not include their containers.

File: src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h (hunk 2 above)
Comment (FG-TUM) [reply src/autopas/containers/verletClusterLists/VerletClusterListsRebuilder.h]: 6d823227
